### ISSUE TITLE
Jump Table Commands in Wrong Order

### ISSUE THREAD

**Issue by andrewdunsworth** [martinisgroup/servers#281]
Once like every hundred scans or so the jump table command will happen before the pulse sequence (or visa versa).  I noticed this in my version of spin echo where the second pi/2 pulse would sometimes just jump right next to the pi pulse, but I think this happens during our version of swap spectroscopy as well as there will be random points at long delay times that have a very large population (as if the pi pulse is jumping right next to the readout pulse).  Can someone packet sniff this out? @maffoo @DanielSank @zchen088 @JulianSKelly

**Comment by zchen088** [comment]
Pictorial example:
We want this:
pi/2 ----- pi ------ pi/2 measure
But we get this:
pi/2 ----- pi pi/2 ------ measure

**Comment by andrewdunsworth** [comment]
This was in the SiQ branch under s = ['','SiQ','data', 'Qubit', 'SQSi Pla 16', 'Dry', '151203-Cal3'] to be specific (sq.spin_echo_long).

**Comment by DanielSank** [comment]
Looks like a delay and a pi/2 pulse are switching places?

**Comment by andrewdunsworth** [comment]
yes.

**Comment by maffoo** [comment]
I don't think this is a problem for packet sniffing, at least not yet. The first thing you should do is look at the jump table commands your code is actually generating to send to the servers. If you rerun your code, do you get blips at the same location? (That is, do the same inputs to your scan function reliably generate bad sequences?)

**Comment by zchen088** [comment]
We ran our scan with the same input multiple times and it only occasionally failed. Also, we noticed that within the same run of the experiment, some of the stats would be bad and some of the stats would be good.

Here's a good run:
![screenshot_2015-12-04-15-59-18](https://cloud.githubusercontent.com/assets/10439313/11604387/b1a1bb56-9aa0-11e5-9a9a-182c6fa7d665.png)

which corresponds to:
```Play SRAM - IDLE - Play SRAM - IDLE - Play SRAM```

and here's a bad run with the exact same parameters
![screenshot_2015-12-04-15-59-41](https://cloud.githubusercontent.com/assets/10439313/11604384/9468fca2-9aa0-11e5-8cdb-d0c34f8911f6.png)

which looks more like:
```Play SRAM - Short IDLE - Play SRAM - Play SRAM - ????```

Does this point to more of a hardware and firmware issue?

**Comment by JulianSKelly** [comment]
Given that the software will always be outputting the same thing I think this indicates a hardware failure.

**Comment by maffoo** [comment]
within the same run of the experiment, some of the stats would be bad and some of the stats would be good.

Does this mean that after a single start command, some of the N repetitions output a correct sequence and some an incorrect sequence? That would _definitely_ be an fpga hardware problem.

**Comment by zchen088** [comment]
Yes it appears so :(

**Comment by ejeffrey** [comment]
Could also be SRAM writes happening in the middle of the sequence.

**Comment by pomalley** [comment]
Yes, we've definitely had issues with SRAM writes happening while the
sequence is running, at least when debugging. I don't know that it has been
thoroughly tested in the wild, e.g. with long scans and stuff.

To look at this, you can plug two of the scope lines into the mon1 and mon2
connectors on the back of the board, and then play around with what gets
written to the monitors (i.e. you'd want to look at SRAM start and sequence
start, etc.)

**Comment by pomalley** [comment]
Specifically here is the code you need to change to change the monitors (IIRC, things have been shifted a bit recently):

https://github.com/martinisgroup/servers/blob/master/fpgalib/dac.py#L1205

**Comment by zchen088** [comment]
I'll try to get started debugging, but can we get someone with more GHzDAC-fu to help us next week? This seems important.

**Comment by DanielSank** [comment]
> can we get someone with more GHzDAC-fu to help us next week? This seems important.

Absolutely. I finished the data run on Vince and have nothing major on my todo list. Happy to help. Some things you guys could do:

* Peter's suggestion about looking at the monitors.

* Install Wireshark and @maffoo's plug-in for dissecting our FPGA packets.

* Modify fpgaseqtransmon to log the packets it's preparing for the qubit sequencer and make sure they're right. I think there may already be a flag you can set in the module so that it writes packet data to a file.

**Comment by DanielSank** [comment]
@andrewdunsworth and/or @zchen088  can you summarize what we learned yesterday here?

**Comment by andrewdunsworth** [comment]
When this problem occurs it occurs on every stat of an experiment.  In the case of sq.swap_spectroscopy it seemed to only effect one board (the z board) where instead of running the step_down, jump table idle, and step_up it ran the step_down, step_up (all that was written in SRAM) repeatedly every 16 us. i.e:

intended scan: -----|_________|---------
actual scan:     -----|_|------------|_|----------|_|----...

the readout, xy, and adc boards did not seem to be effected by this problem.

We then learned that the way the DAC boards work is that the daisy chain fires on every stat of an experiment (which wasn't obvious to me).

**Comment by andrewdunsworth** [comment]
actual scan*:  ---------|_|----------|_|----------|_|-----...

**Comment by DanielSank** [comment]
@andrewdunsworth if you put backticks around some text it shows up as a fixed with font. You can also fence a block of code with three back ticks.

**Comment by andrewdunsworth** [comment]
ok it's getting rid of my underscores in the middle but you get the idea.

**Comment by maffoo** [comment]
Ok, so problem occurs on every repetition of a given experiment. Does it occur every time that particular sequence is run?

**Comment by DanielSank** [comment]
> Ok, so problem occurs on every repetition of a given experiment.

We are not 100% sure of that.

> Does it occur every time that particular sequence is run?

No.

**Comment by andrewdunsworth** [comment]
No, it happens seemingly randomly about every 10 to 100 times a given scan is run.

**Comment by andrewdunsworth** [comment]
And only if jump table commands are called (we are using idle, nop, and end).

**Comment by DanielSank** [comment]
@andrewdunsworth you can edit your comment posts after they're posted, fyi.

**Comment by andrewdunsworth** [comment]
>No, it happens seemingly randomly about every 10 to 100 times a given scan is run.

**Comment by andrewdunsworth** [comment]
Here are the two sequences as seen on the scope:
![20151209_134003](https://cloud.githubusercontent.com/assets/8405285/11699747/c220c854-9e7a-11e5-83d9-01a9470a003a.jpg)
![20151209_134024](https://cloud.githubusercontent.com/assets/8405285/11699748/c3802bea-9e7a-11e5-95c4-f1dd087f4c86.jpg)

Green is the z-board, blue is the xy, yellow is the ADC demod, and pink is the monitor_0 of the readout dac board which is set to 5 which according to 

https://github.com/martinisgroup/servers/blob/u/maffoo/dac-docs/fpgalib/docs/GHzDAC_v8.md

is the SRAM running:

![image](https://cloud.githubusercontent.com/assets/8405285/11699885/8a8c6cc6-9e7b-11e5-92b6-17b0252a5f9a.png)

the pulse on the dac monitor seems to happen before and after the sequence when the pulses occur normally, but are no where to be found when the pulses occur incorrectly.

**Comment by andrewdunsworth** [comment]
When we put the adc_1 monitor on the SRAM write it seems like it writes at the same place for both the 'good' pulse sequence and the 'messed up' one:

good:
![20151209_134024](https://cloud.githubusercontent.com/assets/8405285/11700711/3f607fb2-9e80-11e5-9883-945200774dcb.jpg)

messed:
![20151209_142022](https://cloud.githubusercontent.com/assets/8405285/11700728/5ae3b1a0-9e80-11e5-8313-abcb6c9dd4fa.jpg)

**Comment by andrewdunsworth** [comment]
We put blue on monitor_0 = 5 which is the SRAM running option. It seems to know that it is running SRAM the entire time as the pulses look like this:

Good:
![20151209_142022](https://cloud.githubusercontent.com/assets/8405285/11701126/92be4e08-9e82-11e5-98be-d4727241f73d.jpg)
Bad:
![20151209_142022](https://cloud.githubusercontent.com/assets/8405285/11701133/9cd1c438-9e82-11e5-8edb-d17e4270e2ab.jpg)

**Comment by maffoo** [comment]
@andrewdunsworth, do you have packet wireshark packet captures for two runs like these? It would be really helpful to see exactly what got sent to the boards.

**Comment by andrewdunsworth** [comment]
Ok so if we set blue to IDLE and pink to END (I'm guessing these are the jump table commands), it looks like:
good:
![20151209_144249](https://cloud.githubusercontent.com/assets/8405285/11701246/50b045f6-9e83-11e5-8433-90554096a0c4.jpg)
bad:
![20151209_144228](https://cloud.githubusercontent.com/assets/8405285/11701267/63d91ed2-9e83-11e5-9968-4331c1385b0d.jpg)

So it seems to think that it is idling for the entire length of the messed up pulse sequence... which doesn't make sense (notice that there is a slight down tick where the actual idle occurs at the end of the messed up sequence).

**Comment by andrewdunsworth** [comment]
This really makes it look like its a jumptable problem and not a hardware problem.

**Comment by andrewdunsworth** [comment]
We are working on getting wireshark up and running.

**Comment by andrewdunsworth** [comment]
We now have wireshark but are unsure how to parse the data.

**Comment by andrewdunsworth** [comment]
We are now sharking wires.
